tycho-client/src/feed/component_tracker.rs

```diff
@@ -128,16 +133,60 @@ where
             .into_iter()
             .map(|pc| (pc.id.clone(), pc))
             .collect::<HashMap<_, _>>();
-        self.update_contracts();
+
+        self.initialise_contracts();
+
         Ok(())
     }
 
-    fn update_contracts(&mut self) {
-        self.contracts.extend(
-            self.components
-                .values()
-                .flat_map(|comp| comp.contract_ids.iter().cloned()),
-        );
+    /// Initialise the tracked contracts list from tracked components and their entrypoints
+    fn initialise_contracts(&mut self) {
+        // Add contracts from all tracked components
+        self.contracts = self
+            .components
+            .values()
+            .flat_map(|comp| comp.contract_ids.iter().cloned())
+            .collect();
+
+        // Add contracts from entrypoints that are linked to tracked components
+        let tracked_component_ids = self
+            .components
+            .keys()
+            .cloned()
+            .collect::<HashSet<_>>();
+        for (_, (components, contracts)) in self.entrypoints.iter() {
+            if !components.is_disjoint(&tracked_component_ids) {
+                self.contracts
+                    .extend(contracts.iter().cloned());
+            }
+        }
+    }
+
+    /// Update the tracked contracts list with contracts associated with the given components
+    fn update_contracts(&mut self, components: Vec<String>) {
+        // Only process components that are actually being tracked. Convert to HashSet for
+        // efficient lookup.
+        let tracked_component_ids = components
+            .iter()
+            .filter(|&id| self.components.contains_key(id))
+            .map(|s| s.to_string())
+            .collect::<HashSet<_>>();
```

Comment:
We also can avoid these clones no? With something like this (might need to slightly adapt the filter)

```suggestion
        // Only process components that are actually being tracked. Convert to HashSet for
        // efficient lookup.
        let tracked_component_ids = components
            .into_iter()
            .filter(|&id| self.components.contains_key(id))
            .collect::<HashSet<_>>();
```